service/history/workflow/update/abort_reason_test.go

```diff
@@ -0,0 +1,48 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package update
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAbortReasonUpdateStateMatrix(t *testing.T) {
+	for r := AbortReasonRegistryCleared; r < lastAbortReason; r++ {
+		for st := stateCreated; st < lastState; st <<= 1 {
+			fe, ok := reasonStateMatrix[reasonState{r: r, st: st}]
+			// If new abort reason or state is added, this test will fail.
+			// Do not modify the test but make sure to update the reasonStateMatrix.
+			require.True(t, ok, "Missing combination: %v, %v. If new abort reason or state is added make sure to update the reasonStateMatrix", r, st)
+			if fe.f != nil {
+				require.Nil(t, fe.err)
+			}
+			if fe.err != nil {
+				require.Nil(t, fe.f)
+			}
+		}
+	}
+}
```

Comment:
Should there be a test to ensure that `f` and `err` can not both be non-nil?

Reply:
There are such cases for `stateCompleted` and `stateAborted`. They are not actually used (there is a check on caller side) but I would like to keep them there to make this matrix complete.

Reply:
Ah; I missed that! 👍